031-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,168 @@
+This enables users to access the REST API with TLS if required.
+The `KafkaConnector`, `KafkaConnect` and `KafkaMirrorMaker2` operators will use the TLS encrypted listener when it is enabled, even if an unencrypted listener is present.
+
+The `restListeners` field is designed for potential future extension, and is a list of items that contains a `protocol` field and an optional `restCertChainAndKey` field.
+
+For this proposal, the `protocol` field only supports two values: `http` and `https`. If the `restListeners` list is empty the `KafkaConnect` and `KafkaMirrorMaker2` 
+runtimes will be created with a single unencrypted listener on 8083 to match the existing behaviour.
+
+The `restCertChainAndKey` field will be required if the `https` protocol is selected.
+
+For example, the following CR will enable the HTTP listener on port 8083 and the HTTPS listener on port 8443:
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+spec:
+  restListeners:
+  - protocol: http
+  - protocol: https
+  ...
+```
+
+The schema for `restListeners` will be as follows:
+```
+openAPIV3Schema:
+  type: object
+  properties:
+    spec:
+      type: object
+      properties:
+        restListeners:
+          type: array
+          items:
+            type: object
+            properties:
+              protocol:
+                type: string
+                enum:
+                - http
+                - https
+                description: The protocol of the REST listener.
+              restCertChainAndKey:
+                type: object
+                properties:
+                  certificate:
+                    type: string
+                    description: The name of the ca certificate file in the Secret.
+                  key:
+                    type: string
+                    description: The name of the ca private key file in the Secret.
+                  secretName:
+                    type: string
+                    description: The name of the Secret containing the certificate.
+                  required:
+                    - key
+                    - certificate
+                    - secretName
+                description: CA certificate to use to generate.
+            required:
+            - protocol
+          description: List of additional REST listeners.
+        ...
+```
+
+the above CR would result in the following properties in the generated connect configuration:
+
+```
+listeners: https://:8443,http://:8083
+rest.advertised.listener: https
+rest.advertised.port: 8443
+listeners.https.ssl.client.auth: none
+listeners.https.ssl.truststore.location: /tmp/kafka/kafka-connect-rest.truststore.p12
+listeners.https.ssl.truststore.password: ***generated password***
+listeners.https.ssl.truststore.type: PKCS12
```

Comment:
Wait ... what is the truststore based on? How do users generate their user certificates for this?

Reply:
These properties were copied from the original proposal, but you're right, if this is only hosting a HTTPS endpoint and not doing mutual TLS would we only need a keystore, not a truststore?

Reply:
Yes, correct. It would need only the keystore. And the operator will only need to trust the CA / public key to be able to connect.